### PR TITLE
fix(connection): reliable BLE/manual reconnect after reboot — keep GATT link through reboot, drive CLI-save reconnects

### PR DESCRIPTION
**Pull request by haslinghuis**
Fixes the Bluetooth reconnect-after-reboot regression introduced with #5209 — and the underlying cause of "deaf" BLE reconnects (missing `stopNotifications()` on teardown) that also broke plain disconnect → connect — plus manual-mode (WebSocket) reconnect hardening and a dead BT-11 CRC bypass found during the audit. Bench-verified on a SpeedyBee F405 Mini over Web Bluetooth on Linux (BlueZ): connect, disconnect → connect, and Save & Reboot all reconnect reliably. Also exercised against a SITL build through a ws proxy.

## Root cause (four stacked defects)

1. **CLI-save flows never reconnect BLE.** #5209 changed `scheduleReconnect()` (presets apply, backups restore) to drop the stale link and rely on auto-connect picking up the re-enumerated device. That works for USB serial, but BLE and manual/TCP links never emit `removedDevice`/`addedDevice` for an FC reboot, so nothing ever reconnected.
2. **Reconnected GATT sessions are deaf on Linux/BlueZ.** The reboot flow dropped the surviving GATT link at the flush timeout and reconnected. On Linux (where Web Bluetooth is experimental) a rapid `gatt.disconnect()`/`connect()` cycle reliably produces a session that opens and enumerates services but never delivers notifications — every MSP handshake timed out, on every retry, for the whole window.
3. **One dead handshake consumed the whole reboot window.** The retry loop skips ticks while a connection is open, and an opened-but-silent link was held for the generic 10 s `connecting` watchdog — whose teardown also runs `setArmingEnabled`, an MSP round-trip a silent FC can never answer.
4. **Stale watchdogs killed healthy successors.** `GUI.timeout_add` does not de-duplicate names, and the unexpected-disconnect path never cleared the dead connection's `connecting`/`connectAttempt` timers, so attempt N's watchdog could tear down attempt N+1.

## Fixes

- New `serial_backend.scheduleRebootReconnect()`: drives the flush → reset → retry cycle for reboots initiated outside the module (CLI `save`/`exit`), without re-sending `MSP_SET_REBOOT`. `useMspCliSession.scheduleReconnect()` routes `bluetooth*`/`manual` targets to it; serial keeps the passive re-enumeration path.
- **Keep the BLE link open through the reboot** (`softResetForReboot`): when Auto-Connect will retry, only the app-level connection state is reset and the retry rides the existing GATT session — `gatt.connect()` is a no-op on a connected device, and the bound-once listeners from #5209 make the services/notifications re-setup idempotent. Hardware that genuinely drops the link falls back to the previous hard-reconnect path. If the reboot window closes without a live handshake, `releaseKeptRebootLink()` drops the kept transport for real so it cannot linger behind a "disconnected" UI.
- Reboot-aware handshake stall: during the retry loop an opened-but-silent link is dropped after 3 s (soft on a kept link, hard on a re-established one) and fed back to the loop; driven reboots get a 20 s window (serial stays at 10 s, dialog tracks the same window).
- `finishUnexpectedDisconnect` clears the dead connection's watchdogs; `abortConnection` gates its failure dialog on the live retry loop so mid-reboot failures stay silent.
- Diagnostics: disconnects are now attributed in the console (unsolicited GATT drop vs flush vs intentional), and WebBluetooth connection-setup failures log to the console instead of only `gui_log`.

## Manual-mode audit follow-ups

- **WebSocket superseded-socket guard**: each `connect()` reassigned `this.ws`, but an abandoned attempt's late `onclose` ran `disconnect()` against — and closed — the *current* socket, and its `disconnect` event fed `onClosed` against the wrong session. `connect()` now detaches and closes any previous socket up front, and every handler no-ops once its socket is superseded — including after `onmessage`'s async blob decode, so a socket replaced mid-decode cannot leak stale bytes into the new session.
- **Failed opens signal `connect:false`**: a `WebSocket`-constructor rejection (e.g. a raw `tcp://` manual override, which browsers cannot open — raw TCP is the desktop app's `TauriTcp` transport) previously died silently until the pre-open watchdog; it now dispatches `connect:false` so the connect flow recovers immediately.
- **Revived the dead BT-11/CC2541 CRC bypass**: `msp.js` gated it on `serial.protocol === "bluetooth"`, but that getter returns the lowercased constructor name (`"webbluetooth"`/`"capacitorble"`), so the bypass never ran. It is now a capability check on the active protocol (`shouldBypassCrc`), which only the Bluetooth protocols implement.
- Tauri paths verified: `TauriSerial` opens arbitrary OS paths (`/dev/rfcomm0` manual overrides) and `TauriTcp` dispatches loop-compatible `connect:false`/`disconnect` events — no changes needed.

## Deaf-session root cause: notification teardown

`disconnect()` never called `stopNotifications()`. Chrome tracks an active notify session per characteristic; without an explicit stop before the GATT link drops, `startNotifications()` on the NEXT connection can resolve on Linux/BlueZ without re-arming the CCCD — a session that looks subscribed but never receives. This is what made every re-established session deaf (and plain disconnect → connect fail even after the reboot flow was fixed). Teardown now stops notifications while the link is still up (skipped on unplug/unsolicited drops; failures are logged and do not abort cleanup). The keep-the-link reboot strategy remains: it avoids the reconnect entirely and is the fastest path back to a working session.

## Review follow-ups (CodeRabbit + adversarial review)

- WebSocket `onmessage` re-checks the superseded-socket guard after the async decode boundary; a message decoded across a supersede is dropped.
- `rebootLinkKept` is re-derived from protocol truth (`serial.connected`) at each retry tick: a real transport close landing while serial_backend's disconnect listener is detached (between attempts) would otherwise leave the flag stale, and a freshly re-established — possibly deaf — session would be soft-ridden as the known-good kept one.
- WebBluetooth `connect()` serializes behind the published teardown promise (teardown is genuinely async now that it awaits `stopNotifications()`), and the completion check is optional-chained — an unsolicited teardown mid-connect completes as a signaled `connect:false` instead of a swallowed TypeError that burned the attempt.
- `releaseKeptRebootLink()` skips while an attempt is in flight (`connecting_to` set) — disconnecting under a live `connect()` coroutine would tear the device out from under it; the attempt's own watchdog chain owns that cleanup.

## Reboot-window ownership + progress-based stall (adversarial review)

- The reboot window (start + duration) now lives in `ConnectionState` as a single snapshot taken at `requestReboot(windowMs)`. The retry loop, the reboot dialog, and `abortConnection`'s dialog suppression all read that one source instead of re-deriving from the live port picker each tick — removing the module-level `rebootTimestamp`, the per-tick reactive read, and any loop/dialog deadline divergence if the port selection changes mid-window. The loop also stops when the window's other owner concludes it (a closed window reads as not-expired), and a bug where the dialog check-timer and the loop each nulled the shared window and left the other spinning was fixed.
- Handshake stall is now progress-based: any bytes received in the last stall slice grant another slice instead of restarting the whole handshake, so a slow-but-answering BLE bridge no longer livelocks against the 3 s cap — the reboot window still bounds total time.
- A single `isDrivenRebootTarget(port)` helper replaces the bluetooth/manual predicate that had been spelled out four times across `serial_backend` and `useMspCliSession`.

## Stuck blank-content-area fix

An unstable BLE link fires a burst of disconnect events during a reboot reconnect, each running `teardownConnectionUi()`. After the first teardown lands on the landing tab, every later one called `unmountVueTab()` (blanking the content) then `switchTab("landing")`, which no-ops because `GUI.active_tab` is already `"landing"` — leaving the content area unmounted and stuck black. A racing unmount could also null `activeTabName` under an in-flight mount and leave `tab_switch_in_progress` stuck true, blocking all later tab switches.

- `switchTab()` now dedups only when the target is the active tab **and** actually mounted (`vueTabState.activeTabName === tabKey`), so it remounts a tab left unmounted instead of refusing.
- `teardownConnectionUi()` only unmounts when actually leaving the current tab; when already on the destination it just (re)mounts via `switchTab`, so repeated teardowns never blank the screen or race an in-flight mount.

## Reboot dialog must not conclude on a stale connection

A BLE/manual link survives the reboot command (only the MCU restarts), so `CONFIGURATOR.connectionValid` stays stale-true until the flush drops it ~1.5s later. The reboot dialog's 100ms check-timer concludes on `connectionValid`, and since the reconnect window now lives in `ConnectionState`, concluding nulls it — so the retry loop's first tick reads `!isRebootWindowOpen`, treats it as timed-out, and stops before ever reconnecting. `reinitializeConnection` now resets `CONFIGURATOR.connectionValid = false` right after sending `MSP_SET_REBOOT` (restoring a guard that existed in the old `gui.js` reinitializeConnection and was dropped in the `serial_backend` migration), so the dialog waits for a real reconnect. Verified on hardware: the loop arms, retries, reconnects, and concludes on the fresh `connectionValid`.

## Also in this PR (from the same investigation)

- **Manual mode**: `tcp://` manual overrides can't be opened by a browser (raw TCP is the desktop app's `TauriTcp` transport) — a `WebSocket`-constructor rejection now signals `connect:false` instead of dying silently.
- **BT-11/CC2541 CRC bypass revived**: `msp.js` gated it on `serial.protocol === "bluetooth"`, but that getter returns the constructor name (`"webbluetooth"`), so it was dead — now a capability check on the active protocol.
- **i18n**: added the missing `configurationSaved` message key (pre-existing `missingKey` console warning).

## Testing

- Bench: SpeedyBee F405 Mini, Chrome on Linux — connect, disconnect → connect, and Save & Reboot (Configuration tab, Auto-Connect on) all reconnect reliably. Before these changes, five consecutive re-established sessions were GATT-healthy but MSP-deaf.
- SITL (raw TCP behind a ws proxy, the web-shell setup): drove the real `Websocket` protocol end-to-end — connect, `MSP_API_VERSION` round-trip (API 1.48), then the reboot-loop pattern: disconnect → reconnect → MSP still flows.
- 377 unit tests pass. Coverage added across the investigation: `connectionValid` reset on reboot, repeated-disconnect teardown not blanking the content area, kept-link flag re-derivation after an unheard drop, window expiry leaving in-flight attempts alone, teardown/connect serialization + signaled failure on mid-connect teardown, progress-based stall (slow-but-answering link kept, silent one dropped), notification teardown (stop on live link / skip on dead link / failure-tolerant), kept-link soft reset + re-handshake, 20 s driven window, kept-link release on window expiry, no failure dialogs mid-loop, watchdog cleanup on unexpected disconnect, `scheduleRebootReconnect`, `scheduleReconnect` transport routing, WebSocket superseded-socket guard (incl. across the async decode) + failed-open signaling, and the CRC-bypass capability gate.

## Remaining serial-backend work — recommended order (issue #4722)

This PR closes out the connection-state part of the serial-backend cleanup tracked in #4722. The four still-open items there were re-verified against the current code and sequenced **low-risk → high-risk**, keeping any wide mechanical change out of the transport paths this PR just stabilized. Full detail and evidence live on the issue; the order is:

1. **`bytes.js` conversion helper** (item 4 · S · low) — extract only the base64↔`Uint8Array` pair duplicated byte-for-byte in `CapacitorTcp.js` + `CapacitorBle.js` (the issue's "~8 files" is really 2), add a round-trip test. Self-contained, touches no transport state machine.
2. **`port → device` rename** (item 5 · L · med) — one isolated, behaviour-neutral PR. 290 sites in `src/` + ~71 in tests (ship together). Don't touch genuine OS-port identifiers or the Tauri plugin `path` param; decide `getConnectedPort → getConnectedDevice` deliberately.
3. **`SerialProtocolBase`** (item 3 · M · high) — **defer + downgrade.** `connect/disconnect/send` share signatures, not bodies; only ~30 trivial lines are hoistable, and the blast radius lands on the reconnect/GATT paths this PR and #5209 just fixed. If done, a trivial-helper mixin only, gated on green serial tests.
4. **`path → protocol`** (item 6) — **drop/close.** Collides with the existing `protocol` transport-type concept; `path` is routing-significant and bound to the Tauri `serialplugin` native param.

**Order:** `bytes.js` → `port→device` rename → `SerialProtocolBase` (deferred) → close `path→protocol`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Improved reboot reconnect workflow for Bluetooth/manual connections with a configurable reconnect window, refined retry/auto-connect behavior, and consistent reboot dialog timing.
  * Added “Configuration saved” status text.
* **Bug Fixes**
  * Prevented stale/superseded connection actions in WebSocket/WebBluetooth with safer teardown and notification stop handling.
  * Improved reconnect/teardown edge cases by clearing stale reconnect/handshake timers and handling unexpected disconnects reliably.
  * Fixed MSP CRC-bypass decision logic for supported corrupted-message cases.
* **Tests**
  * Expanded reconnect, teardown, superseded-socket, reboot timing, and reconnect routing coverage.

### CODE REVIEW

**Review by nerdCopter — APPROVED** [review]

File: src/js/protocols/WebBluetooth.js [hunk 1]
```diff
@@ -211,11 +226,19 @@ class WebBluetooth extends EventTarget {
...
         }
 
-        // Bluetooth API doesn't provide a way for getInfo() or similar to get the connection info
-        const connectionInfo = this.device.gatt.connected;
+        // Bluetooth API doesn't provide a way for getInfo() or similar to get the connection info.
+        // Optional-chained: an unsolicited disconnect during the setup above runs the full
+        // teardown (a kept session's connected=true bypasses the cancel branch) and nulls
+        // this.device — the attempt must then complete as a signaled failed open
+        // (connect:false below), not die on a TypeError that serial.connect swallows.
+        const connectionInfo = this.device?.gatt?.connected;
```
[the remark this answers is not in the document]
Comment (haslinghuis) [reply src/js/protocols/WebBluetooth.js]: Addressed in 15f083f6 — `connect()` now tracks a `setupComplete` flag and gates the success branch on it. A throw in `getServices`/`getCharacteristics`/`startNotifications` after `gattConnect()` now dispatches `connect:false` instead of a truthy `connect` for a characteristic-less session.

File: src/js/serial_backend.js [hunk 2]
```diff
@@ -232,10 +275,51 @@ function beginDisconnect() {
...
+
+// BLE variant of disconnectForReboot: keep the GATT session OPEN through the reboot, reset
+// only app-level state, and let the retry loop ride the existing session. Dropping and
+// re-establishing the link is the fragile part — on Linux/BlueZ a rapid disconnect/connect
+// cycle yields a "deaf" session (opens and enumerates, but no notifications arrive).
+function softResetForReboot() {
+    console.log(`${logHead} Keeping BLE link through reboot — app-level reset only (flush timeout)`);
+    prepareDisconnect();
+    resetAppConnectionState();
+    rebootLinkKept = true;
```
[the remark this answers is not in the document]
Comment (haslinghuis) [reply src/js/serial_backend.js:325]: Addressed in 15f083f6 — `softResetForReboot()` now clears `intentionalDisconnect` after `prepareDisconnect()`, since the kept-link path triggers no close to consume it.

File: test/js/webbluetooth.test.js [hunk 3]
```diff
@@ -265,6 +265,37 @@ describe("WebBluetooth openCanceled abort contract", () => {
...
+        bt.devices = [bt.createPort(device)];
+        bt.device = device;
+        const characteristic = makeReadCharacteristic();
+        bt.readCharacteristic = characteristic;
+
+        await bt.disconnect();
+
+        expect(characteristic.stopNotifications).toHaveBeenCalledTimes(1);
+        expect(bt.readCharacteristic).toBe(false); // teardown completed
+        expect(device.gatt.disconnect).toHaveBeenCalled();
```
[the remark this answers is not in the document]
Comment (haslinghuis) [reply test/js/webbluetooth.test.js:295]: Addressed in 15f083f6 — the test now asserts `stopNotifications` runs before `gatt.disconnect` via `invocationCallOrder`, pinning the ordering that is the deaf-session regression.

File: test/js/websocket.test.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+describe("Websocket protocol — superseded socket guard (manual/SITL reconnect)", () => {
+    beforeEach(() => {
+        vi.stubGlobal("WebSocket", FakeWebSocket);
+    });
+
+    afterEach(() => {
+        vi.unstubAllGlobals();
+    });
```
[the remark this answers is not in the document]
Comment (haslinghuis) [reply test/js/websocket.test.js:25]: Addressed in 15f083f6 — added `vi.restoreAllMocks()` in `afterEach` so a `console` spy can't leak into later specs.

File: src/js/serial_backend.js [hunk 5]
```diff
@@ -1167,15 +1353,31 @@ function rebootReconnect() {
...
+            const state = getConnectionState();
+            const timedOut = state.rebootWindowExpired || !state.isRebootWindowOpen;
             if (CONFIGURATOR.connectionValid || timedOut || !PortHandler.portPicker.autoConnect) {
                 stopRebootReconnect();
                 // The reboot window has closed (reconnected, timed out, or auto-connect off):
-                // concludeReboot settles to IDLE so normal selection resumes.
+                // concludeReboot settles to IDLE so normal selection resumes. A kept BLE
+                // link that never made it back to connected is dropped for real here.
                 getConnectionState().concludeReboot(CONFIGURATOR.connectionValid);
+                releaseKeptRebootLink();
```
[the remark this answers is not in the document]
Comment (haslinghuis) [reply src/js/serial_backend.js:1371]: Skipping this one — deliberate, not an oversight.

Traced the concrete outcome: if Auto-Connect is toggled off *while a reconnect attempt is already in flight*, the retry loop stops but that in-flight attempt completes → the user ends up **connected**. That's a consistent, usable state — not a stuck or broken one. The reboot window is concluded (`concludeReboot`), the phase settles to `CONNECTED`, `linkOpen` is true.

Gating the connect success path (or aborting a mid-flight GATT open) touches the delicate `connect()`/`openCanceled` interaction that took several rounds to stabilise and was just verified on real hardware (SpeedyBee F405 Mini). Re-entering that path for a benign "connected slightly more than the user asked" edge carries real regression risk for no user-visible correctness gain.

@coderabbitai if you can construct a sequence where this leaves a genuinely *broken* state (stuck spinner, half-open transport, phantom connection) — rather than just "connected after toggle-off" — please share the exact steps and I'll reconsider.

File: src/js/serial_backend.js [hunk 6]
```diff
@@ -530,8 +682,13 @@ function abortConnection(messageKey) {
...
     // Connection failed dialog appears even though it connected" report). Gate on auto-connect:
     // without it nothing retries, so a failure there is real and must still surface. Capture
-    // before setPhase(FAILED) overwrites the reconnect phase.
-    const duringRebootReconnect = getConnectionState().isRebootReconnecting && PortHandler.portPicker.autoConnect;
+    // before setPhase(FAILED) overwrites the reconnect phase. The open reboot window is
+    // checked too: after the loop's FIRST failed attempt the phase has already left
+    // REBOOTING/RECONNECTING (onOpen/abort advanced it), so the phase alone would let
+    // attempt 2+'s failures pop the dialog mid-reboot.
+    const duringRebootReconnect =
+        (getConnectionState().isRebootReconnecting || getConnectionState().isRebootWindowOpen) &&
```
Comment (blckmn) [inline src/js/serial_backend.js]: A leaked window makes this suppress real failure dialogs forever. The serial path with suppressDialog=true skips showRebootDialog and nothing else concludes the window. No caller passes true today, so not blocking, but worth closing off. Either conclude on that branch or also check rebootWindowExpired here.

File: test/js/serial_backend.test.js [hunk 7]
```diff
@@ -653,123 +644,3 @@ describe("serial_backend removedDevice matching is device-specific", () => {
...
 });
-
-// ---------------------------------------------------------------------------
-// Characterization — pins CURRENT reboot behavior for the serial/USB path
-// and the virtualMode path of reinitializeConnection(). These assert what the
-// code does TODAY (not the desired end-state) so later refactor slices have a
-// regression net. If a later slice intentionally changes this behavior, these
-// tests are expected to be updated alongside that change.
-// ---------------------------------------------------------------------------
-describe("serial_backend reinitializeConnection — serial/USB reboot path (characterization)", () => {
```
Comment (blckmn) [inline test/js/serial_backend.test.js:664]: These two blocks still look valid and were the only net over the serial/USB and virtual reboot paths, which this PR also touches. What made them unviable? If it's just the connectionValid reset, my suggestion is updating them instead of dropping the coverage.

**Review by blckmn — APPROVED** [review]
A fair few of the new comments tell the investigation story rather than the constraint. The PR keeps the history, the code doesn't need to. Worth a trim.

File: src/js/serial_backend.js (hunk 6 above)
Comment (haslinghuis) [reply src/js/serial_backend.js]: Good catch — closed both ways in b6d5eb34. The serial path now concludes the reboot window when the dialog is suppressed (nothing else would), and `abortConnection`'s suppression is additionally gated on `!rebootWindowExpired` so even a leaked window stops hiding real failures once its duration elapses.

File: test/js/serial_backend.test.js (hunk 7 above)
Comment (haslinghuis) [reply test/js/serial_backend.test.js:664]: Agreed — restored in b6d5eb34. They were dropped in an aggressive test trim, but you're right they were the only net over the serial/USB and virtual reboot paths this PR touches. Brought them back reframed as behaviour guards (dropped the "characterization / pins current behaviour" framing) and updated to also assert the new `connectionValid = false` reset on the serial path.

**Comment by haslinghuis** [comment]
Thanks for the end-to-end walk, @blckmn. Both inline points addressed in b6d5eb34 (replies inline).

On the comment style: agreed — the code shouldn't carry the investigation narrative. Trimmed the worst offenders to state the constraint instead (the PR history keeps the story). If any others still read as story rather than constraint, point them out and I'll cut them too.

File: src/js/msp.js [hunk 8]
```diff
@@ -281,7 +281,12 @@ const MSP = {
         if (this.message_checksum === expectedChecksum) {
             // message received, store dataview
             this.dataView = new DataView(this.message_buffer, 0, this.message_length_expected);
-        } else if (serial.protocol === "bluetooth" && serial._protocol?.shouldBypassCrc?.(expectedChecksum)) {
+        } else if (serial._protocol?.shouldBypassCrc?.(expectedChecksum)) {
+            // Capability check: only the Bluetooth protocols (WebBluetooth, CapacitorBle)
+            // implement shouldBypassCrc, for BT-11/CC2541 bridges that corrupt the MSP
+            // checksum to 0xff. Deliberately NOT gated on `serial.protocol === "bluetooth"`:
+            // that getter returns the lowercased constructor name ("webbluetooth"/
+            // "capacitorble"), so the old comparison never matched and the bypass was dead.
```
Comment (blckmn) [inline src/js/msp.js:289]:
```suggestion
            // Capability check: only the Bluetooth protocols implement shouldBypassCrc,
            // for BT-11/CC2541 bridges that corrupt the MSP checksum to 0xff. Not gated
            // on serial.protocol — that getter returns the lowercased constructor name,
            // never "bluetooth".
```

File: src/js/protocols/WebBluetooth.js [hunk 9]
```diff
@@ -85,6 +88,9 @@ class WebBluetooth extends EventTarget {
         if (this.closeRequested) {
             return;
         }
+        // Explicitly attribute the teardown: this is the DEVICE dropping the link
+        // (or the FC rebooting under it), not the configurator closing it.
```
Comment (blckmn) [inline src/js/protocols/WebBluetooth.js:92]:
```suggestion
```

File: src/js/protocols/WebBluetooth.js [hunk 10]
```diff
@@ -191,6 +197,15 @@ class WebBluetooth extends EventTarget {
     }
 
     async connect(path, options) {
+        // Serialize behind an in-flight teardown: doCleanup awaits stopNotifications, so
+        // disconnect() is genuinely async — a connect that started mid-teardown would
+        // reset closeRequested under it and then have its device/characteristics torn
+        // down by the resumed cleanup. The teardown promise never rejects.
```
Comment (blckmn) [inline src/js/protocols/WebBluetooth.js:203]:
```suggestion
        // Serialize behind an in-flight teardown: a connect started mid-teardown would
        // reset closeRequested under it and have its device/characteristics torn down
        // by the resumed cleanup. The teardown promise never rejects.
```

File: src/js/protocols/WebBluetooth.js [hunk 11]
```diff
@@ -210,12 +226,22 @@ class WebBluetooth extends EventTarget {
             await this.getServices();
             await this.getCharacteristics();
             await this.startNotifications();
+            setupComplete = true;
         } catch (error) {
+            // Also log to the console — gui_log only reaches the in-app log panel, which
+            // makes a failed services/characteristics/notifications setup invisible in
+            // console captures while the connection still reports "opened".
```
Comment (blckmn) [inline src/js/protocols/WebBluetooth.js:233]:
```suggestion
```

File: src/js/protocols/WebBluetooth.js [hunk 12]
```diff
@@ -210,12 +226,22 @@ class WebBluetooth extends EventTarget {
...
         }
 
-        // Bluetooth API doesn't provide a way for getInfo() or similar to get the connection info
-        const connectionInfo = this.device.gatt.connected;
+        // Require setup to have COMPLETED, not just gatt.connected. If getServices/
+        // getCharacteristics/startNotifications threw, gatt can still report connected
+        // while the session has no usable characteristics — treating that as success
+        // dispatches connect:true for a link that would only ever time out on MSP.
+        // Optional-chained: an unsolicited disconnect during setup nulls this.device, so
+        // the attempt must complete as a signaled failed open, not a swallowed TypeError.
```
Comment (blckmn) [inline src/js/protocols/WebBluetooth.js:243]:
```suggestion
        // gatt can report connected while setup threw and left no usable characteristics,
        // so require setup to have completed. Optional-chained: an unsolicited disconnect
        // during setup nulls this.device.
```

File: src/js/protocols/WebBluetooth.js [hunk 13]
```diff
@@ -373,6 +406,20 @@ class WebBluetooth extends EventTarget {
                 // before calling, or false.removeEventListener throws and aborts cleanup.
                 if (this.readCharacteristic) {
                     this.readCharacteristic.removeEventListener("characteristicvaluechanged", this.handleNotification);
+
+                    // Explicitly stop notifications BEFORE dropping the GATT link (only
+                    // possible while it is still up — an unplug/unsolicited drop skips
+                    // this). Without it, Chrome keeps the notify session marked active,
+                    // and on Linux/BlueZ startNotifications() on the NEXT connection can
+                    // resolve without re-arming the CCCD — a session that looks
+                    // subscribed but never receives (deaf), breaking disconnect/connect.
```
Comment (blckmn) [inline src/js/protocols/WebBluetooth.js:415]:
```suggestion
                    // Stop notifications while the link is still up (an unplug/unsolicited
                    // drop skips this): Chrome keeps the notify session marked active, and
                    // on Linux/BlueZ the next connection's startNotifications() can then
                    // resolve without re-arming the CCCD — subscribed but never receiving.
```

File: src/js/protocols/WebSocket.js [hunk 14]
```diff
@@ -61,10 +61,41 @@ class Websocket extends EventTarget {
...
+        // has been replaced by a newer attempt, otherwise a stale onclose would run
+        // disconnect() against — and close — the newer socket.
+        let ws;
+        try {
+            ws = new WebSocket(this.address, ["binary", "wsSerial"]);
+        } catch (e) {
+            // Invalid URL/scheme — e.g. a raw tcp:// manual override, which a browser
+            // cannot open (raw TCP needs the desktop app's native transport). Signal a
+            // failed open so the connect flow recovers immediately instead of dying
+            // silently until the pre-open watchdog.
```
Comment (blckmn) [inline src/js/protocols/WebSocket.js:87]:
```suggestion
            // Invalid URL/scheme, e.g. a raw tcp:// manual override, which a browser
            // cannot open (raw TCP needs the desktop app's native transport).
```

File: src/js/serial_backend.js [hunk 15]
```diff
@@ -407,22 +496,27 @@ function teardownConnectionUi() {
...
-    } else {
-        switchTab("landing", { mode: "disconnected" });
+    const target = pendingTab === "firmware_flasher" ? "firmware_flasher" : "landing";
+
+    // Only blank the content area when we're actually LEAVING the current tab. When a
+    // burst of disconnects (an unstable BLE link during a reboot reconnect) runs this
+    // teardown repeatedly, we are already on the destination tab: unmounting would blank
+    // the screen while the follow-up switchTab no-ops (same tab) or races an in-flight
+    // mount, leaving a stuck black screen. switchTab remounts if the tab was somehow left
+    // unmounted, and no-ops when it is already showing.
```
Comment (blckmn) [inline src/js/serial_backend.js:510]:
```suggestion
    // Only unmount when actually leaving the current tab: a repeated teardown (disconnect
    // burst) already sits on the destination, and unmounting there blanks the content
    // while switchTab no-ops on the same tab or races an in-flight mount.
```

File: src/js/serial_backend.js [hunk 16]
```diff
@@ -473,22 +572,72 @@ function finishUnexpectedDisconnect() {
     teardownConnectionUi();
 }
 
+// Drop a connection whose MSP handshake stalled during the reboot retry loop, WITHOUT
+// ending the loop: we connected to a still-booting FC, so just close the dead link and let
+// the loop's next tick try again. Deliberately not disconnectForReboot() — its
+// prepareDisconnect() would stopRebootReconnect() and kill the very loop that must retry.
```
Comment (blckmn) [inline src/js/serial_backend.js:578]:
```suggestion
// Drop a connection whose MSP handshake stalled during the reboot retry loop, without
// ending the loop: the FC is still booting, so close the dead link and let the next tick
// try again. Not disconnectForReboot() — its prepareDisconnect() would stopRebootReconnect().
```

File: src/js/serial_backend.js [hunk 17]
```diff
@@ -473,22 +572,72 @@ function finishUnexpectedDisconnect() {
...
+    // onOpen advanced the phase to HANDSHAKING; re-assert the reconnect phase so
+    // notifyClosed() leaves the window open and selectActivePort() keeps aiming at the
+    // rebooting device instead of falling back mid-retry.
+    getConnectionState().reconnectStarted();
+
+    // On a kept BLE link the stall just means the FC hasn't finished booting: keep riding
+    // the known-good session (app-level reset only) and let the next tick re-handshake.
+    // Dropping it here would force a real reconnect — the deaf-session trap this whole
+    // keep-the-link strategy exists to avoid. NOT prepareDisconnect(): its
+    // stopRebootReconnect would kill the live retry interval.
```
Comment (blckmn) [inline src/js/serial_backend.js:589]:
```suggestion
    // On a kept BLE link, keep riding the known-good session (app-level reset only) and
    // let the next tick re-handshake — dropping it would force the deaf-session reconnect
    // the kept link exists to avoid. Not prepareDisconnect(): its stopRebootReconnect
    // would kill the live retry interval.
```

File: src/js/serial_backend.js [hunk 18]
```diff
@@ -1099,7 +1254,15 @@ export function reinitializeConnection(suppressDialog = false) {
     // Send reboot command to the flight controller
     MSP.send_message(MSPCodes.MSP_SET_REBOOT, false, false);
 
-    if (currentPort.startsWith("bluetooth") || currentPort === "manual") {
+    // Force the connection invalid now so the reboot dialog and retry loop wait for a REAL
+    // reconnect. A BLE/manual link survives the reboot command (only the MCU restarts), so
+    // connectionValid stays stale-true until the flush drops it ~1.5s later — without this,
+    // the dialog's 100ms check-timer sees it true immediately, concludes the reboot, and
+    // (since the window moved into ConnectionState) nulls the reconnect window before the
+    // retry loop even arms, so no reconnect ever runs.
```
Comment (blckmn) [inline src/js/serial_backend.js:1262]:
```suggestion
    // A BLE/manual link survives the reboot command (only the MCU restarts), so
    // connectionValid stays stale-true until the flush drops it ~1.5s later. Force it
    // false now so the reboot dialog and retry loop wait for a real reconnect.
```

File: src/js/tab_switch.js [hunk 19]
```diff
@@ -45,7 +45,13 @@ export function switchTab(tabKey, options = {}) {
     const mode = options.mode ?? "disconnected";
     const label = options.label ?? defaultLabel(tabKey);
 
-    if (GUI.active_tab === tabKey || GUI.tab_switch_in_progress) {
+    // Dedup only when the target is BOTH the active tab AND actually mounted. After
+    // unmountVueTab() (disconnect teardown) the content area is blank while GUI.active_tab
+    // still names the old tab, so an active_tab-only check would refuse to remount and
+    // leave a stuck blank screen — the burst of disconnects an unstable BLE link produces
+    // during a reboot reconnect hits this teardown->landing path repeatedly.
```
Comment (blckmn) [inline src/js/tab_switch.js:52]:
```suggestion
    // Dedup only when the target is both the active tab and actually mounted: after
    // unmountVueTab() the content area is blank while GUI.active_tab still names the
    // old tab, and refusing to remount would leave it blank.
```

**Review by blckmn — APPROVED** [review]
Can't push to the branch, so the comment trim is in the suggestions. Batch-apply should do it.
